src/lib.rs

```diff
@@ -1,14 +1,165 @@
+    /// and a boolean indicating whether the hash is from the left (true) or right (false) sibling.
+    pub fn generate_proof(&self, element: &str) -> Option<Vec<(String, bool)>> {
+        if self.leaves.is_empty() {
+            return None;
+        }
+
+        let mut hash = Self::hash(element);
+        let mut proof = Vec::new();
+        let mut current_level = self.leaves.clone();
+
+        while current_level.len() > 1 {
+            let mut next_level = Vec::new();
+            let mut found = false;
+
+            for i in (0..current_level.len()).step_by(2) {
+                let left = &current_level[i];
+                let right = if i + 1 < current_level.len() {
+                    &current_level[i + 1]
+                } else {
+                    left
+                };
+                let parent_hash = Self::hash(&(left.clone() + right));
+                if left == &hash || right == &hash {
+                    proof.push(if left == &hash {
+                        (right.clone(), false)
+                    } else {
+                        (left.clone(), true)
+                    });
+                    hash.clone_from(&parent_hash);
+                    found = true;
+                }
+                next_level.push(parent_hash);
+            }
+
+            if !found {
+                return None;
+            }
+
+            current_level = next_level;
+        }
+
+        Some(proof)
+    }
+
+    /// Verifies that an element is part of the Merkle Tree using the given proof.
+    ///
+    /// # Arguments
+    ///
+    /// * `element` - A string slice representing the element to be verified.
+    /// * `proof` - A vector of tuples representing the proof, where each tuple contains a hash
+    ///   and a boolean indicating whether the hash is from the left (true) or right (false) sibling.
+    ///
+    /// # Returns
+    ///
+    /// A boolean indicating whether the element is part of the tree.
+    pub fn verify(&self, element: &str, proof: Vec<(String, bool)>) -> bool {
+        if let Some(root) = &self.root {
+            let mut hash = Self::hash(element);
+            for (p, is_left) in proof {
+                hash = if is_left {
+                    Self::hash(&(p + &hash))
+                } else {
+                    Self::hash(&(hash + &p))
+                };
+            }
+            hash == *root
+        } else {
+            false
+        }
+    }
 
-    #[test]
-    fn it_works() {
-        let result = add(2, 2);
-        assert_eq!(result, 4);
+    /// Computes the SHA256 hash of the given data.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - A string slice representing the data to be hashed.
+    ///
+    /// # Returns
+    ///
+    /// A string representing the hexadecimal encoding of the SHA256 hash of the data.
+    fn hash(data: &str) -> String {
+        let mut hasher = Sha256::new();
+        hasher.update(data);
+        let result = hasher.finalize();
```

Comment:
This can be simplified as:

```suggestion
        let result = Sha256::digest(data);
```

as showcased in [the crate repository's README](https://github.com/RustCrypto/hashes/tree/master/sha2#one-shot-api).

Reply:
You're right, I'll commit your suggestion.